### DISCUSSION TITLE
How to replace tray menu items with tauri 2.0?

### DISCUSSION THREAD

**Discussion by don41382** [tauri-apps/tauri#11134] [Q&A]
I am trying to display a timer in my tray menu, so it shows the current time.

I thought I just have to replace the whole menu with `tray.set_menu`. This works fine under MacOS, but not under Windows. Under windows the menu only get updated every once in a while and is not consistent.

Is there maybe a different way?

Maybe someone has an idea.

**Comment by Asdmir786** [comment]
Yeah, this is a common Windows quirk.

On macOS, replacing the tray menu with `set_menu` tends to feel live. On Windows, the tray menu is basically a normal native context menu: it’s not meant to be rebuilt every second, so frequent `tray.set_menu(...)` calls often look lagged / inconsistent.

What usually works better in Tauri v2:

1. **Don’t rebuild the whole menu every tick**  
   Keep a `MenuItem` handle in app state (clone it — they’re ref-counted) and update only that label:

```rust
// when building the menu once
let clock = MenuItem::with_id(app, "clock", "12:00:00", false, None::<&str>)?;
app.manage(ClockItem(clock.clone()));

// later, on a timer
if let Some(state) = app.try_state::<ClockItem>() {
    let _ = state.0.set_text(now_string);
}
```

2. **Prefer tooltip for a live clock on Windows**  
   Tray **title** isn’t supported on Windows. A live `set_tooltip("12:34:56")` is usually much more reliable than stuffing a ticking clock into the menu.

3. **If it must be in the menu**  
   Update the stored item with `set_text`, and only call `set_menu` when the *structure* changes (items added/removed, enabled/disabled, etc.). Also avoid updating while the menu is open if you can — Windows often snapshots the menu when it’s shown.

In HalalDL I only refresh the tray menu on real state changes (active downloads, pause, updates, etc.), not on a 1 Hz timer — that pattern has been much happier on Windows.

Hope that helps!
